### PR TITLE
[api-extractor] Add a new setting "tsconfigFilePath"

### PR DESCRIPTION
**Pull request by octogonz**
Fixes issue https://github.com/Microsoft/web-build-tools/issues/1222

This PR adds a new config file setting `tsconfigFilePath` for customizing the path for `tsconfig.json`.

It also improves some config file related error messages.

### CODE REVIEW

File: apps/api-extractor/src/api/ExtractorConfig.ts [hunk 1]
```diff
@@ -494,17 +512,29 @@ export class ExtractorConfig {
...
+
+      const tsconfigFilePath: string = ExtractorConfig._resolvePathWithTokens('tsconfigFilePath',
+        configObject.compiler.tsconfigFilePath, tokenContext);
+
+      if (configObject.compiler.overrideTsconfig === undefined) {
+        if (!tsconfigFilePath) {
+          throw new Error('Either the "tsconfigFilePath" or "overrideTsconfig" setting must be specified');
+        }
+        if (!FileSystem.exists(tsconfigFilePath)) {
+          throw new Error('The "tsconfigFilePath" path does not exist: ' + tsconfigFilePath);
```
Comment (iclanton) [inline apps/api-extractor/src/api/ExtractorConfig.ts]: *The file referenced by "tsconfigFilePath" does not exist: ...

**Comment by octogonz** [comment]
In this code:

```ts
        task.stderr.on('data', (data: string) => {
          writer.writeError(data);
          this._hasWarningOrError = true;
          // If this error matches any of the suppressed warnings from the config, write it to
          // stdout instead of stderr and don't treat the task as having a warning/error.
          if (this._rushConfiguration.suppressedWarnings.some((warning: string) => data.indexOf(warning) !== -1)) {
            writer.write(data);
          } else {
            writer.writeError(data);
            this._hasWarningOrError = true;
          }
        });
```

...I'm not sure that the concept of a "warning" or "line" exists at all.  The `data` string is a chunk of buffered text.  As far as I know, the string may contain the second half of a line, followed by a newline character, and then the first half of the subsequent line.  It may be purely coincidence that certain popular binaries or scripts choose to flush the stream after writing each line of text.

Also, the text buffer sometimes contains color escape characters, and sometimes not.  There are probably other similar technical issues that a generic stream parser may need to consider.

Up until now, Rush has carefully avoided getting involved with stream parsing.  We do collect stderr and interpret it as "there were some warnings" and print it as a summary at the end, but otherwise Rush is pretty agnostic about the tasks that it orchestrates.

I wonder if there is a different approach.  For example, maybe the "filtering bad messages from a tasks's output" could be made into a separate tool, and then the build task could pipe itself through that tool.  This would allow the solution to be more versatile, and developed by a larger audience, and it would give more fine-grained control over which filters are applied to which processes.

On the other hand, we could also just go and fix the bad task that's doing deprecated stuff.  (It may be easier than you estimated.)

Just something to think about...

File: apps/api-extractor/src/api/ExtractorConfig.ts (hunk 1 above)
Comment (octogonz) [reply apps/api-extractor/src/api/ExtractorConfig.ts]: Fixed

**Comment by octogonz** [comment]
This feature was published with the API Extractor beta release 7.0.41

**Pull request merged**